2022Swerve/src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -132,12 +124,51 @@ public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTr
         return DriveConstants.kDriveKinematics.toSwerveModuleStates(chassisSpeeds);  
     }
 
+    private void normalizeDrive(SwerveModuleState[] desiredStates, ChassisSpeeds speeds){
+        double translationalK = Math.hypot(speeds.vxMetersPerSecond, speeds.vyMetersPerSecond) / DriveConstants.kPhysicalMaxSpeed;
+        double rotationalK = Math.abs(speeds.omegaRadiansPerSecond) / DriveConstants.kPhysicalMaxAngularSpeed;
+        double k = Math.max(translationalK, rotationalK);
+      
+        // Find the how fast the fastest spinning drive motor is spinning                                       
+        double realMaxSpeed = 0.0;
+        for (SwerveModuleState moduleState : desiredStates) {
+          realMaxSpeed = Math.max(realMaxSpeed, Math.abs(moduleState.speedMetersPerSecond));
+        }
+      
+        double scale = Math.min(k * DriveConstants.kPhysicalMaxSpeed / realMaxSpeed, 1);
+        for (SwerveModuleState moduleState : desiredStates) {
+          moduleState.speedMetersPerSecond *= scale;
+        }
+    }
+
     public void setModuleStates(SwerveModuleState[] desiredStates) {
-        SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, DriveConstants.kPhysicalMaxSpeed);
+        this.normalizeDrive(desiredStates, DriveConstants.kDriveKinematics.toChassisSpeeds(desiredStates));
         frontLeft.setDesiredState(desiredStates[0]);
         frontRight.setDesiredState(desiredStates[1]);
         backLeft.setDesiredState(desiredStates[2]);
         backRight.setDesiredState(desiredStates[3]);
     }
 
+    public SwerveModuleState[] getModuleStates() {
+        SwerveModuleState[] swerveModuleArray = new SwerveModuleState[4];
+        swerveModuleArray[0] = frontLeft.getState();
+        swerveModuleArray[1] = frontRight.getState();
+        swerveModuleArray[2] = backLeft.getState();
+        swerveModuleArray[3] = backRight.getState();
+        return swerveModuleArray;
+    }
+
+    public void stopModules(){
+        frontLeft.stop();
+        frontRight.stop();
+        backLeft.stop();
+        backRight.stop();
+    }
+
+    @Override
+    public void periodic() {
+        odometer.update(getRotation2d(), getModuleStates());
+        SmartDashboard.putNumber("Robot Heading", getHeading());
```

Comment:
Is this going to be a debug smartdashboard thing?  Or competition dashboard?

We should talk about it with the team, and see whether we can build our plan for disabling debug smartDashboard values during competition.  

We can't decide until we find out whether it affects performance, however I think we should make a plan for how much high speed logging we should disable in competition.  

If we disable Robot Heading in smartdashboard, could we log it separately in high speed logging?

Do we have a logging class that optionally puts to smartdashboard, or to high speed logging?  Let's make a plan, and put a plan into trello cards before we move forward.

( And if the plan is easy, let's start the plan here...if not, we can wait until we work on the card. )

Reply:
> Is this going to be a debug smartdashboard thing? Or competition dashboard?

The odometer update is separate from the dashboard. We don't know what we want/need for the competition dashboard at all in this point of time, but these current dashboard elements are important for working with the initial hardware.

> If we disable Robot Heading in smartdashboard, could we log it separately in high speed logging?

Of course we could do this, or we could even log it simultaneously. But the purpose of these elements is to see what the robot is doing in the real world, and see if that matches what the robot thinks it's doing.

> Do we have a logging class that optionally puts to smartdashboard, or to high speed logging? Let's make a plan, and put a plan into trello cards before we move forward.

No, we do not. If that's something you think we should have, we should put it in a trello card for a later PR.

Reply:
ok agree that we should create a trello card for it.  

> Of course we could do this, or we could even log it simultaneously. But the purpose of these elements is to see what the robot is doing in the real world, and see if that matches what the robot thinks it's doing.

Will it be necessary to see the robot heading realtime during competition? 
Will robot heading be something that we want to disable entirely during competition?
Will robot heading be useful in logging if something during the competition breaks and we need to do a postmortem?

It isn't necessary to do both types of logging, because SmartDashboard uses network tables, and those are logged.
But if we think that Robot Heading ( for example ) is something that we won't care about being on a dashboard during competition because we can see the heading...and we will likely disable it, then we might want to log it instead.

Reply:
Quoting from discord message so everyone can see

> I have some ideas about how this would look.  It would be cool to do logMagic(COMPETITION, "Prefix", "name",value) and have Prefix be the tab name in shuffleboard, or part of the path of the high speed log

I like this idea, we should create a card to have this as a follow on feature. 

For now, the dashboard will stay until we have somewhere else to put it. It's important for testing, and competition dashboard isn't something we know yet.

Reply:
https://trello.com/c/5BfWTk7m